Support component tsl2591

wjcarpenter opened the pull request:
# What does this implement/fix? 

This adds support for the AMS TSL2591 ambient light sensor, which is the replacement for the TSL2561

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [*] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>
https://github.com/esphome/feature-requests/issues/267

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1370

## Test Environment

- [ ] ESP32
- [*] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
i2c:
  # ...
spi:
  # ...
sensor:
  - platform: tsl2591
    id: "my_tls2591"
    name: "TSL2591 Ambient Light"
    address: 0x29
    update_interval: 60s
    visible:
      name: "TSL2591 visible light"
    infrared:
      name: "TSL2591 infrared light"
    full_spectrum:
      name: "TSL2591 full spectrum light"
    calculated_lux:
      name: "TSL2591 Lux"
```

## Checklist:
  - [*] The code change is tested and works locally.
  - [*] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [*] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
working, but with redefined enums

Commit:
tidying up

Commit:
this-ify things according to ESPhome C++ coding guidelines

Commit:
add sensor calculated_lux and clarify other sensors

Commit:
PR lint fixes

Commit:
more PR lint fixes

wjcarpenter commented:
FYI, I changed a file under nextion due to a complaint from one of the PR CI checks. I defined CONF_VISIBLE in consts.py, but nextion had it defined locally.

Commit:
even more PR lint

Commit:
PR lint

Commit:
PR lint

Commit:
PR lint

Commit:
see if I can stymie clang-tidy complaint

Commit:
NOLINT causes a lint error; sheesh

Commit:
another permutation for the adafruit header

Commit:
add Adafruit dependency to PIO so clang-tidy can find the header

Commit:
PR lint

Commit:
I decided to change the default icon for the sensors to mdi:birghtness-6

Commit:
typo in previous commit

oxan commented on esphome/components/tsl2591/sensor.py:
```diff
@@ -0,0 +1,155 @@
+    cg.add(var.set_integration_time(config[CONF_INTEGRATION_TIME]))
+    cg.add(var.set_gain(config[CONF_GAIN]))
+    # https://platformio.org/lib/show/463/Adafruit%20TSL2591%20Library
+    cg.add_library("Adafruit TSL2591 Library", "^1.4.0")
```
```suggestion
    cg.add_library("adafruit/Adafruit TSL2591 Library", "^1.4.0")
```

PlatformIO wants libraries to have an owner nowadays :)

oxan commented on esphome/components/tsl2591/sensor.py:
```diff
@@ -0,0 +1,155 @@
+
+# This enum is a clone of the enum in the Adafruit library.
+# I couldn't work out how to use the Adafruit enum directly in codegen.
+TSL2591IntegrationTime = tsl2591_ns.enum("TSL2591IntegrationTime")
```
```suggestion
TSL2591IntegrationTime = cg.global_ns.enum("tsl2591IntegrationTime_t")
```

This should allow you to get rid of the duplication.

wjcarpenter replied:
moot

oxan commented on esphome/components/tsl2591/sensor.py:
```diff
@@ -0,0 +1,155 @@
+
+# This enum is a clone of the enum in the Adafruit library.
+# I couldn't work out how to use the Adafruit enum directly in codegen.
+TSL2591Gain = tsl2591_ns.enum("TSL2591Gain")
```
```suggestion
TSL2591Gain = cg.global_ns.enum("tsl2591Gain_t")
```

(Note that the Adafruit constants have different names than what you're using below).

wjcarpenter replied:
moot

oxan commented on platformio.ini:
```diff
@@ -18,6 +18,7 @@ lib_deps =
     1655@1.0.2  ; TinyGPSPlus (has name conflict)
     6865@1.0.0  ; TM1651 Battery Display
     6306@1.0.3  ; HM3301
+    463@1.4.0 ; Adafruit TSL2591 Library
```
```suggestion
    adafruit/Adafruit TSL2591 Library@1.4.0
```

The IDs are only necessary if there are duplicate libraries on the PlatformIO registry.

oxan commented on esphome/components/tsl2591/tsl2591.cpp:
```diff
@@ -0,0 +1,124 @@
+
+void TSL2591Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up TSL2591...");
+  if (!this->tsl2591_.begin()) {
```
You should pass the configured address here, so that it works if the user provides a non-default address.

wjcarpenter replied:
Oh, yeah, that's my mistake. Thanks.

I'm actually a little conflicted about making the address configurable at all. The spec sheet for the device says it's 0x29 (and also uses 0x28) and can't be changed. So, you're not going to have more than one of these devices on a single I2C bus. I guess with a mux or some other kind of hardware in the middle you could change the address, so that justifies the configurability.

oxan commented on esphome/components/tsl2591/sensor.py:
```diff
@@ -0,0 +1,155 @@
+
+# The Adafruit sensors library requires both i2c and spi in ESPhome configs, but
+# we don't use spi for this component.
+DEPENDENCIES = ["i2c", "spi"]
```
Are you sure about? It'd be unfortunate to require everyone using this sensor to waste a bunch of pins to an SPI bus that isn't used.

wjcarpenter replied:
It's definite that ESPHome won't accept the config without an `spi:` block. However, it doesn't have to use any real pins. For example, in my config I have:
```
spi:
  clk_pin: 0
  mosi_pin: 0
```

oxan reviewed (commented):
Thanks for your contribution!

Aside from the (minor) comments I've left, I see one issue: ESPHome has its own I2C abstraction layer (which provides e.g. support for multiplexers). I'm not sure using the global `Wire` instance, as the Adafruit library does is compatible with that. Maybe someone a bit more experienced with that than me can give input on that issue.

wjcarpenter committed:
Update esphome/components/tsl2591/sensor.py

Co-authored-by: Oxan van Leeuwen <oxan@oxanvanleeuwen.nl>

wjcarpenter committed:
Update platformio.ini

Co-authored-by: Oxan van Leeuwen <oxan@oxanvanleeuwen.nl>

wjcarpenter commented:
When I first started working on this, I thought it was going to be pretty trivial to just go with the Adafruit library, so I didn't bother with the ESPHome I2C abstraction. In retrospect, it took a lot more time than I had planned (something something software development :-))

I will give the ESPHome I2C stuff another look to see how much bother it would be to change over. As with any I2C device, there's a certain amount of register and bit fiddling, but I can benefit from the Adafruit code that already works.

Commit:
Reimplement TSL2591 using ESPHome I2C instead of the Adafruit library

This removes the dependency on the Adafruit TSL2591 library and the transitive
dependency on having an SPI configuration in ESPHome.

Along the way, I noticed that there was a rather large delay() in the
sensor readings in the Adafruit library. That's no problem for a simple
arduino-like framework, but it's a problem for ESPHome's update loop
for pollable components.

The root cause of the delay was allowing time for the ADCs to have
valid values after being enabled. And that came about because the device
was disabled after a reading and re-enabled for the next reading. That's
pretty good for power-saving, but still bad for ESPHome's update loop.

I made multiple integration changes because of that delay:

1. By default, I don't disable the device at all, so the ADC
readings don't need that delay. The ADCs operate continuously,
and you get the most recent completed integration cycle readings.

2. There is a configurable power saving mode. That gives the
disable/enable behavior that the Adafruit library has, but at the
cost of wonking out the ESPHome update loop. Docs recommend against
using the power saving mode, and it's off by default.

3. I exposed some additional C++ methods so that someone could
script their way through the enable-delay-read-disable process. Delays
in scripts are async, so they're not a problem for the ESPHome update
loop. (I have not tested doing this scripting.)

Commit:
PR formatting lint

Commit:
PR non-functional formatting lint

Commit:
PR non-functional formatting lint

Commit:
PR non-functional formatting lint

wjcarpenter commented:
I took @oxan's suggestion and reimplemented this using ESPHome I2C, as reflected in the current PR code. That caused some ripple effect changes because I discovered a big delay loop in the Adafruit library. You can see details of that in my commit comments for https://github.com/esphome/esphome/pull/2131/commits/babb9a389e1e6f1730c45c4d927846a979818a4c

oxan commented on esphome/components/tsl2591/tsl2591.cpp:
```diff
@@ -0,0 +1,295 @@
+void TSL2591Component::disable_internal_() {
+  if (this->power_save_mode_enabled_) {
+    this->disable();
+  }
```
Given that `disable()` already checks if the power save mode is enabled, this method seems unnecessary?

If that's the case, it might be nice to make it clear from the name of `disable` that it only disables if powersaving is on, so e.g. `disable_if_power_saving()` / `enter_power_saving()` / `device_idle()` / something like that.

wjcarpenter replied:
Just a careless mistake. I copied the method but neglected to remove the check from the original.

oxan commented on esphome/components/tsl2591/tsl2591.cpp:
```diff
@@ -0,0 +1,295 @@
+  ESP_LOGCONFIG(TAG, "Setting up TSL2591...");
+  auto id = this->read_byte(TSL2591_COMMAND_BIT | TSL2591_REGISTER_DEVICE_ID);
+  if (id != 0x50) {
+    ESP_LOGD("ERROR", "Could not find a TSL2591 Sensor. Did you configure I2C with the correct address?");
```
It might be useful to log the `id` result here as well. In the past (for other components) we've had users with a counterfeit/fake chip that had a different device id. Logging the `id` will make it easier to diagnose such issues.

wjcarpenter replied:
Good suggestion. When I went to make that change, I realized that none of the logging from my setup() is showing up anywhere. (I also changed logging the error to ESP_LOGE.) Maybe it's because I'm testing with OTA? Or maybe it's something related to the start-up sequence and it doesn't show up for anybody?

I'm going to leave the logging in there in case it's something environmental that prevents it showing up for me

wjcarpenter replied:
Since i2c.cpp logs I2C problems with ESP_LOGI, I guess I'll do the same thing for the chip ID mismatch.

oxan replied:
Yes, errors during setup() are logged before the network connection is made, so if you look at the logs through wifi, you won't see them. They're present on the serial bus though.

~You should log the error using `ESP_LOGE` (the last letter is the severity of the message, E = error, W = warn, I = info, D = debug).~ (ah, i see you already mentioned that). You should also use the `TAG` constant as the first parameter, that labels the log message with the integration.

wjcarpenter replied:
Agreed. I copied this section from some other integration. It will now be ESP_LOGE(TAG, ...

oxan commented on esphome/components/tsl2591/tsl2591.cpp:
```diff
@@ -0,0 +1,295 @@
+  TSL2591IntegrationTime raw_timing = this->integration_time_;
+  int timing_ms = (1 + raw_timing) * 100;
+  ESP_LOGCONFIG(TAG, "  Integration Time: %d ms", timing_ms);
+  ESP_LOGCONFIG(TAG, "  Power save mode enabled: %s", this->power_save_mode_enabled_ ? "True" : "False");
```
```suggestion
  ESP_LOGCONFIG(TAG, "  Power save mode enabled: %s", ONOFF(this->power_save_mode_enabled_));
```
For consistency with what other components use :)

oxan commented on esphome/components/tsl2591/tsl2591.cpp:
```diff
@@ -0,0 +1,295 @@
+  this->enable();
+  this->integration_time_ = integration_time;
+  this->write_byte(TSL2591_COMMAND_BIT | TSL2591_REGISTER_CONTROL, this->integration_time_ | this->gain_);  // NOLINT
+  disable_internal_();
```
This method gets called very early in startup, before e.g. the I2C multiplexer is setup, so you shouldn't do communication here. Instead, only set the `integration_time_` variable here, and actually write that command to the device in `setup()`. That also avoids a redundant enable/disable cycle for setting both integration time and gain.

wjcarpenter replied:
I didn't know that, but I'm making this change. Since the actual I2C transaction sets them both into the same register, I've changed the exposed API to require both params. A nice side effect is that I can do the power management thing inside this combined setter method to avoid bogus ADC values on the next read.

oxan commented on esphome/components/tsl2591/tsl2591.cpp:
```diff
@@ -0,0 +1,295 @@
+  this->enable();
+  this->gain_ = gain;
+  this->write_byte(TSL2591_COMMAND_BIT | TSL2591_REGISTER_CONTROL, this->integration_time_ | this->gain_);  // NOLINT
+  this->disable_internal_();
```
Same here.

oxan commented on esphome/components/tsl2591/tsl2591.cpp:
```diff
@@ -0,0 +1,295 @@
+
+uint32_t TSL2591Component::get_combined_illuminance() {
+  this->enable();
+  // Wait x ms for ADC to complete and signal valid.
```
If I understand correctly, this delay will only happen if powersaving mode is enabled, right? It would be nice to add that to the comment.

wjcarpenter replied:
good suggestion

oxan commented on esphome/components/tsl2591/tsl2591.h:
```diff
@@ -0,0 +1,205 @@
+///
+/// This is a fairly thin ESPHome compatibility wrapper around the
+/// Adafruit TSL2591 library. Adafruit uses the term "luminosity", but
+/// ESPHome uses the term "illuminance", as does the TLS2591 datasheet.
```
This comment is now obsolete.

oxan commented on esphome/components/tsl2591/tsl2591.h:
```diff
@@ -0,0 +1,205 @@
+   * This gets an individual light sensor reading. Since it goes through
+   * the entire component read cycle to get one value, it's not optimal if
+   * you want to get all possible channel values. If you want that, first
+   * call `getCombinedIlluminance()` and pass that value to the companion
```
```suggestion
   * call `get_combined_illuminance()` and pass that value to the companion
```

oxan commented on esphome/components/tsl2591/tsl2591.h:
```diff
@@ -0,0 +1,205 @@
+   * It should take no more than the configured integration time for
+   * the ADC values to become valid after the TSL2591 device is enabled.
+   */
+  boolean is_adc_valid();
```
```suggestion
  bool is_adc_valid();
```
`boolean` is an Arduino-ism.

wjcarpenter replied:
what's an arduino? :-)

oxan reviewed (changes requested):
Thanks! Overall this looks fairly good now, just some relatively minor cleanups left :)

oxan commented:
Btw, it's also possible to implement the safe-delay scripting you talk about in the component directly. It's fine to leave that out for now (you can do a follow-up PR if you're interested), but for future reference I'll detail the implementation here:

* Move the logic to read from the sensors to a separate method (e.g. `process_update()`)`
* In `update`, if the ADC is valid, call `process_update()` and return.
* If the ADC isn't valid, set an interval using `set_interval()`. 
* In the interval function, check if the ADC is valid, if so, cancel the interval and call `process_update()`.
* You can use either a static variable or `millis()` to check how often the interval was invoked and ensure it doesn't run forever.

wjcarpenter committed:
Update esphome/components/tsl2591/tsl2591.cpp

Co-authored-by: Oxan van Leeuwen <oxan@oxanvanleeuwen.nl>

wjcarpenter committed:
Update esphome/components/tsl2591/tsl2591.h

Co-authored-by: Oxan van Leeuwen <oxan@oxanvanleeuwen.nl>

wjcarpenter committed:
Update esphome/components/tsl2591/tsl2591.h

Co-authored-by: Oxan van Leeuwen <oxan@oxanvanleeuwen.nl>

wjcarpenter committed:
Merge branch 'esphome:dev' into support-component-tsl2591

Commit:
code review feedback, and allow setting device and glass attentuation factors

Commit:
PR lint

Commit:
PR lint

Commit:
PR lint

Commit:
PR lint

wjcarpenter commented:
Great suggestion about using intervals, @oxan. I much prefer taking care of it by default rather than having to bother each user to think about it. I implemented that and changed the default for power_save_mode to be enabled.

BTW, I really appreciate the effort you are putting into reviewing the code. You have an eye for detail and an attitude for quality. (And, somehow, you are able to read my mind and know when I am being lazy or careless. :-) )

Commit:
implement delay for ADC using intervals

Commit:
PR lint

Commit:
PR lint

Commit:
PR lint

Commit:
PR lint

Commit:
PR lint

wjcarpenter committed:
Merge branch 'esphome:dev' into support-component-tsl2591

oxan commented on esphome/components/tsl2591/tsl2591.cpp:
```diff
@@ -0,0 +1,369 @@
+  this->status_clear_warning();
+}
+
+#define interval_name "tsl2591_interval_for_update"
```
Intervals are namespaced to components, so the prefix isn't really necessary, but it doesn't hurt either :)

wjcarpenter replied:
Are they also scoped to the component instance? If I have two instances of this class, will they both try to use the interval with that fixed name?

wjcarpenter replied:
I had a look at the interval stuff in Component.cpp. I see that it is keyed by component instance, so there would be no interference from multiple instances of a component class.

oxan commented on esphome/components/tsl2591/tsl2591.h:
```diff
@@ -0,0 +1,245 @@
+  float get_setup_priority() const override;
+
+ protected:
+  const char *name_;  // TODO: extend esphome::Nameable
```
Meh, I wouldn't do that, as `Nameable` has grown beyond just being a name and has all sorts of irrelevant stuff now. You technically don't even need a name for this component (only sensors need to be named), though I can see why it's useful if you've got multiple sensors connected.

wjcarpenter replied:
It just seemed a little ragged to me to require names for sensors and not even allow them for the component. (I first tried to use inheritance from Nameable, but ran into some problems.) I am more used to a model where things have an machine-friendly ID that is immutable after creation and a human-friendly name that is mutable.

Anyhow, it didn't cost much :-).

oxan reviewed (approved):
Nice work! Looks good now :tada: I appreciate the extensive comments you've included :)

OttoWinter reviewed (approved):
Thanks!

jesserockz committed:
Merge branch 'dev' into support-component-tsl2591
